pkg/kube/cluster-init.sh

```diff
@@ -70,17 +103,85 @@ get_default_intf_IP_prefix() {
         awk -v new_ip="$ip_prefix" '{gsub("IPAddressReplaceMe", new_ip)}1' /etc/multus-daemonset.yaml > /tmp/multus-daemonset.yaml
 }
 
+# kubernetes's name must be lower case and '-' instead of '_'
+convert_to_k8s_compatible() {
+        echo "$1" | tr '[:upper:]_' '[:lower:]-'
+}
+
+wait_for_device_name() {
+        logmsg "Waiting for DeviceName from controller..."
+        EdgeNodeInfoPath="/persist/status/zedagent/EdgeNodeInfo/global.json"
+        while true; do
+                if [ -f $EdgeNodeInfoPath ]; then
+                        dName=$(jq -r '.DeviceName' $EdgeNodeInfoPath)
+                        if [ -n "$dName" ]; then
+                                HOSTNAME=$(convert_to_k8s_compatible "$dName")
+                                break
+                        fi
+                fi
+                sleep 5
+        done
+
+        # we should have the uuid since we got the device name
+        DEVUUID=$(/bin/hostname)
+        # get last 5 bytes of the DEVUUID as suffix to the hostname
+        DEVUUID_HASH=$(echo "$DEVUUID" | tail -c 6)
+        HOSTNAME="$HOSTNAME-$DEVUUID_HASH"
```

Comment:
Why do we need a name of this form?  Can't we just use the devuuid as the hostname in k3s?

Reply:
we used to have just the k3s node name as device-uuid, but it's hard to debug and in particular in cluster node, it's much nice to see the meaningful names in debugging. we have converted all the PODs, VMIs and nodes names to this format.

Reply:
But then why do we need a few bytes of the uuid?
The edge node name (DeviceName) is unique across a controller.

(Maybe the full device uuid can be part of some additional description field in kubernetes if such a thing exists.)

Reply:
this has been addressed in the broken up pr: https://github.com/lf-edge/eve/pull/3893